### PR TITLE
Progress Reporting for Long Running Operations + Custom Commands

### PR DESCRIPTION
**Pull request by oakeyc**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by troydai** [comment]
Please fix CI first.

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+from enum import Enum
+
+BAR_LEN = 100
+
+
+class ProgressType(Enum):  # pylint: disable=too-few-public-methods
+    """ the types of progress """
+    Determinate = 0
+    Indeterminate = 1
+    Both = 2
```
Comment (brendandburns) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: What is Both mean? Perhaps a more descriptive name or at least a comment to help explain?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    def write(self, args):
+        """ writes the progress """
+        pass
+
+    def flush(self):
+        """ flushes the message out the pipeline"""
+        self.out.flush()
+
+
```
Comment (brendandburns) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: inconsistent spacing (2 here, 1 above)

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+class DetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self, total_value=None):
+        self.message = ''
+        self.curr_val = 0 if total_value else None
+        self.total_val = total_value
+
+    def add(self, args):
+        """ adds a progress report """
```
Comment (brendandburns) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: describe the args?

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 4]
```diff
@@ -187,22 +168,70 @@ def on_input_timeout(self, cli):
...
+        sub_name = ""
+        try:
+            sub_name = PROFILE.get_subscription()[_SUBSCRIPTION_NAME]
+        except CLIError:
+            pass
+
+        curr_cloud = "Cloud: {}".format(get_active_cloud_name())
+        tool_val = '{}'.format('Subscription: {}'.format(sub_name) if sub_name else curr_cloud)
+
+        # progress = get_progress_message()
```
Comment (brendandburns) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: delete dead code.

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 5]
```diff
@@ -393,7 +423,8 @@ def _special_cases(self, text, cmd, outside):
         break_flag = False
         continue_flag = False
 
-        if text and text.split()[0].lower() == 'az':
+        if text and len(text.split()) > 0\
```
Comment (brendandburns) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: I think this should all fit on one line.

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 6]
```diff
@@ -516,7 +547,13 @@ def handle_scoping_input(self, continue_flag, cmd, text):
             return continue_flag, cmd
         return continue_flag, cmd
 
+    def handle_long_process(self, args):
+        """ makes changes to indicates long running process """
+        print('pass')
```
Comment (brendandburns) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: do you just mean `pass`?

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 7]
```diff
@@ -516,7 +547,13 @@ def handle_scoping_input(self, continue_flag, cmd, text):
...
+        """ in LongRUnningOperation there is the call method which could be usefull """
```
Comment (brendandburns) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: nit: LongRunningOperation (note the extra capital 'U')

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py [hunk 8]
```diff
@@ -125,6 +125,31 @@ def reg_extra_arg(self, *args, **kwargs):
 
 
 def _get_client(service, parsed_args):
+    # if hasattr(parsed_args, 'account_name'):
```
Comment (brendandburns) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py]: delete dead code.

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+
+class DeterminateStandardOut(DeterminateProgressView):
+    """ custom output for progress reporting """
+    def __init__(self, out=None):
+        super(DeterminateStandardOut, self).__init__(
+            out if out else sys.stderr, self._format_value)
+
+    def _format_value(self, msg, percent):  # pylint: disable=no-self-use
+        bar_len = BAR_LEN - len(msg) - 1
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: I assume this doesn't work if `bar_len < 0`?
To prevent `bar_len` from going negative, you could truncate `msg` if it's too long. (see `truncate_text() util method in azure.cli.core.util`)

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        completed = int(bar_len*percent)
+
+        message = '\r{}['.format(msg)
+        for i in range(bar_len):
+            if i < completed:
+                message += '#'
+            else:
+                message += ' '
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Don't need this for-loop to go char by char.
see below...
screen shot 2017-05-02 at 10 05 41 am

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        message += ']  {:.4%}'.format(percent)
+        return message
+
+    def write(self, args):
+        """ writes the progress """
+        args = args[0]
+        msg = args['message'] if 'message' in args else ''
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: why not just have keyword arguments for this? Like `message=None, value=..., ...`

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+        if 'value' in args and 'total_val' in args:
+            percent = args["value"] / args['total_val']
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: `from __future__ import division` might be needed for Python 2 division.

File: src/azure-cli-core/tests/test_progress.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import unittest
+import mock
+
+import azure.cli.core.commands.progress as progress
```
Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: `from azure.cli.core.commands.progress import (ProgressType, etc...)`
Then you don't need to keep doing `progress.X` in code.

File: src/azure-cli-core/tests/test_progress.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    def test_indet_stdview(self):
+        """ tests the indeterminate progress standardout view """
+        view = progress.IndeterminateStandardOut()
+        self.assertEqual(view.get_type, progress.ProgressType.Indeterminate)
+        before = view.spinner.total
+        self.assertEqual(view.spinner.label, 'In Progress')
+        view.write({})
+        after = view.spinner.total
+        self.assertTrue(after > before)
+        view.write({'message':'TESTING'})
```
Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: This is why I'd suggest keyword arguments.
`view.write(message='TESTING')` is a lot clearer than what you have.

File: src/azure-cli-core/tests/test_progress.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        self.assertEqual(view.spinner.label, 'TESTING')
+
+    def test_det_stdview(self):
+        """ test the determinate progress standardout view """
+        outstream = MockOutstream()
+        view = progress.DeterminateStandardOut(out=outstream)
+        self.assertEqual(view.get_type, progress.ProgressType.Determinate)
+        args = {'message':'hihi', 'total_val':1, 'value':.5}
+        view.write({args})
```
Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: This is also why I think keyword args would work best.

File: src/azure-cli-core/tests/test_progress.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        # 95 length, 48 complete, 4 dec percent
+        bar_str = ''
+        for i in range(95):
+            if i < 48:
+                bar_str += '#'
+            else:
+                bar_str += ' '
```
Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: `bar_string = ('#'*48).ljust(95)`

Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: Same with below...

File: src/azure-cli-core/tests/test_progress.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        self.assertEqual(view.flush(), '\rhihi[{}]  50.0000%'.format(bar_str))
+        args = {'message':'', 'total_val':1, 'value':.9}
+        # 99 length, 90 complete, 4 dec percent
+        bar_str = ''
+        for i in range(95):
+            if i < 90:
+                bar_str += '#'
+            else:
+                bar_str += ' '
+        self.assertEqual(view.flush(), '\r[{}]  90.0000%'.format(bar_str))
```
Comment (derekbekoe) [inline src/azure-cli-core/tests/test_progress.py]: Can we show max 2dp? 4 is a bit much.

File: src/azure-cli-core/azure/cli/core/commands/progress.py (hunk 10 above)
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Something like `('#'*completed).ljust(bar_len)`

File: src/azure-cli-nspkg/azure/cli/__init__.py [hunk 18]
```diff
@@ -0,0 +1,2 @@
+import pkg_resources
+pkg_resources.declare_namespace(__name__)
```
Comment (derekbekoe) [inline src/azure-cli-nspkg/azure/cli/__init__.py]: This change shouldn't be checked in unless you've talked to @johanste about this...

File: src/command_modules/azure-cli-shell/azclishell/app.py (hunk 7 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: Also, `usefull` -> `useful`

File: src/command_modules/azure-cli-shell/azclishell/app.py (hunk 6 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: Is this a feature that's not yet implemented?

File: src/command_modules/azure-cli-shell/azclishell/az_completer.py [hunk 19]
```diff
@@ -199,7 +199,8 @@ def gen_dynamic_completions(self, text):
                             for comp in gen_dyn_completion(
                                     comp, started_param, prefix, text):
                                 yield comp
-
+                    # except AttributeError:  # for non-robust completion logic
+                    #     pass
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-shell/azclishell/az_completer.py]: Remove dead code.

File: src/command_modules/azure-cli-shell/setup.py [hunk 20]
```diff
@@ -35,6 +35,7 @@
     'applicationinsights',
     'azure-cli-core',
     'jmespath',
+    'humanfriendly',
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-shell/setup.py]: This should actually be added as a dep. to `azure-cli-core` as that's where it's used, not here.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 21]
```diff
@@ -95,10 +95,14 @@ def __setattr__(self, name, value):
...
 class LongRunningOperation(object):  # pylint: disable=too-few-public-methods
 
-    def __init__(self, start_msg='', finish_msg='', poller_done_interval_ms=1000.0):
+    def __init__(self, start_msg='', finish_msg='',
+                 poller_done_interval_ms=1000.0):
         self.start_msg = start_msg
         self.finish_msg = finish_msg
         self.poller_done_interval_ms = poller_done_interval_ms
+        self.controller = ProgressHook(progress_type=ProgressType.Indeterminate)
+        from azure.cli.core.application import APPLICATION
```
Comment (oakeyc) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: this was a cyclic imports problem

Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: @oakeyc @derekbekoe suggests move it to the head of the method, not the file, does it will cause a cyclic imports issue?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 22]
```diff
@@ -95,10 +95,14 @@ def __setattr__(self, name, value):
 
 class LongRunningOperation(object):  # pylint: disable=too-few-public-methods
 
-    def __init__(self, start_msg='', finish_msg='', poller_done_interval_ms=1000.0):
+    def __init__(self, start_msg='', finish_msg='',
+                 poller_done_interval_ms=1000.0):
         self.start_msg = start_msg
         self.finish_msg = finish_msg
         self.poller_done_interval_ms = poller_done_interval_ms
+        self.controller = ProgressHook(progress_type=ProgressType.Indeterminate)
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: rename: `progress_controller` is a better name.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 23]
```diff
@@ -107,7 +111,11 @@ def __call__(self, poller):
         from msrest.exceptions import ClientException
         logger.info("Starting long running operation '%s'", self.start_msg)
         correlation_message = ''
+        self.controller.begin()
+        percent = 0.0
         while not poller.done():
+            percent += .005
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: is there a reason for .5% incremental? Is this value arbitrary?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 24]
```diff
@@ -122,8 +130,10 @@ def __call__(self, poller):
             except KeyboardInterrupt:
                 logger.error('Long running operation wait cancelled.  %s', correlation_message)
                 raise
+
         try:
             result = poller.result()
+
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: You don't need this blank line.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 25]
```diff
@@ -145,6 +155,7 @@ def __call__(self, poller):
 
         logger.info("Long running operation '%s' completed with result %s",
                     self.start_msg, result)
+        self.controller.end()
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Will any exception cause the progress controller not be ended? What's the consequence?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+
+class ProgressType(Enum):  # pylint: disable=too-few-public-methods
+    """ the types of progress """
+    Determinate = 0
+    Indeterminate = 1
+    # for views that want to accept both determinate and indeterminate progres
+    Both = 2
+
+
+class _ProgressViewBase(object):
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Remove the underline. It is not particularly useful.

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 27]
```diff
@@ -0,0 +1,180 @@
...
+    """ a view base for progress reporting """
+    def __init__(self, out, format_percent=None):
+        super(DeterminateProgressView, self).__init__(out, ProgressType.Determinate, format_percent)
+
+
+class InDeterminateProgressView(_ProgressViewBase):
+    """ a view base for progress reporting """
+    def __init__(self, out, format_percent=None):
+        super(InDeterminateProgressView, self).__init__(
+            out, ProgressType.Indeterminate, format_percent)
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Remove these two classes. They do not define new behaviors. Avoid deep and unnecessary class inheritance.

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+        return {'message': self.message}
+
+
+class ProgressHook(object):
+    """ sends the progress to the view """
+    def __init__(self, progress_type):
+        if progress_type == ProgressType.Determinate:
+            self.reporter = DetProgressReporter()
+        elif progress_type == ProgressType.Indeterminate:
+            self.reporter = InDetProgressReporter()  # pylint: disable=redefined-variable-type
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: `If-Else` here is anti-pattern. What should be done is let the caller of the `__init__` to determine what type of the reporter and send in instance.

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+
+
+class DeterminateStandardOut(DeterminateProgressView):
+    """ custom output for progress reporting """
+    def __init__(self, out=None):
+        super(DeterminateStandardOut, self).__init__(
+            out if out else sys.stderr, DeterminateStandardOut._format_value)
+
+    @staticmethod
+    def _format_value(msg, percent):
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Can this just be a normal function?

File: src/azure-cli-core/tests/test_progress.py [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+        return self.type
+
+    def flush(self):
+        return self.string
+
+
+class TestProgress(unittest.TestCase):  # pylint: disable=too-many-public-methods
+    """ test the progress reporting """
+
+    def test_det_model(self):
```
Comment (troydai) [inline src/azure-cli-core/tests/test_progress.py]: I suggest renaming all these test method names to be more explicit about the module they are testing. eg. `test_progress_indicator_det_mode;`

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 31]
```diff
@@ -187,22 +168,56 @@ def on_input_timeout(self, cli):
...
+        try:
+            sub_name = PROFILE.get_subscription()[_SUBSCRIPTION_NAME]
+        except CLIError:
+            pass
+
+        curr_cloud = "Cloud: {}".format(get_active_cloud_name())
+        tool_val = '{}'.format('Subscription: {}'.format(sub_name) if sub_name else curr_cloud)
+
+        settings_items = [
+            " [F1]Layout",
```
Comment (troydai) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: the space here is intentional?

File: src/command_modules/azure-cli-shell/azclishell/app.py [hunk 32]
```diff
@@ -187,22 +168,56 @@ def on_input_timeout(self, cli):
...
+            empty_space += " "
+
+        settings = self._toolbar_info()
+        settings, empty_space = space_toolbar(settings, cols, empty_space)
+        cli.buffers['bottom_toolbar'].reset(
+            initial_document=Document(u'{}{}{}'.format(NOTIFICATIONS, settings, empty_space)))
+
+        # cli.request_redraw()
+        # if return_val:
+        #     return stop
```
Comment (troydai) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: remove.

File: src/command_modules/azure-cli-shell/azclishell/app.py (hunk 31 above)
Comment (oakeyc) [inline src/command_modules/azure-cli-shell/azclishell/app.py]: Yes

**Comment by derekbekoe** [comment]
@oakeyc please comment when we should take another look.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 33]
```diff
@@ -95,10 +95,18 @@ def __setattr__(self, name, value):
 
 class LongRunningOperation(object):  # pylint: disable=too-few-public-methods
 
-    def __init__(self, start_msg='', finish_msg='', poller_done_interval_ms=1000.0):
+    def __init__(self, start_msg='', finish_msg='',
+                 poller_done_interval_ms=1000.0):
         self.start_msg = start_msg
         self.finish_msg = finish_msg
         self.poller_done_interval_ms = poller_done_interval_ms
+        self.progress_controller = self._init_progress()
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: I would expect to be able to pass in a progress_controller instance when creating LongRunningOperation instance.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 34]
```diff
@@ -145,6 +156,7 @@ def __call__(self, poller):
 
         logger.info("Long running operation '%s' completed with result %s",
                     self.start_msg, result)
+        self.progress_controller.end()
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Is it expected that exceptions cause the end() method to be bypassed?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 35]
```diff
@@ -107,7 +115,9 @@ def __call__(self, poller):
         from msrest.exceptions import ClientException
         logger.info("Starting long running operation '%s'", self.start_msg)
         correlation_message = ''
+        self.progress_controller.begin()
         while not poller.done():
+            self.progress_controller.add(message='Running')
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: We already have a start/end message as part of the long running operation instance. Should we use the start message instead of a hard-coded Running message?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 36]
```diff
@@ -200,6 +212,8 @@ def __init__(self, name, handler, description=None, table_transformer=None,
 
     @staticmethod
     def _should_load_description():
+        from azure.cli.core.application import APPLICATION
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Out of curiosity, why was this change needed?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 37]
```diff
@@ -0,0 +1,169 @@
...
+
+    def write(self, args):
+        """ writes the progress """
+        pass
+
+    def flush(self):
+        """ flushes the message out the pipeline"""
+        self.out.flush()
+
+    def get_type(self):
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Is there a specific reason to wrap the accessor in a method? (It would be more pythonic to access the progress_type attribute directly instead of wrapping it in a method)

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 38]
```diff
@@ -0,0 +1,176 @@
...
+        total_val = kwargs.get('total_val', 1.0)
+        value = kwargs.get('value', 0.0)
+        assert value >= 0 and value <= total_val and total_val >= 0
+        self.total_val = total_val
+        self.curr_val = value
+        self.message = message
+
+    def report(self):
+        """ report the progress """
+        percent = self.curr_val / self.total_val
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: What happens if self.total_val is zero? The assert in add checks that it is greater than or equals to zero, which I assume implies that zero is a valid value? :)

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 39]
```diff
@@ -0,0 +1,176 @@
...
+    def update(self):
+        """ updates the view with the progress """
+        for view in self.active_progress:
+            view.write(self.reporter.report())
+            view.flush()
+
+    def stop(self):
+        """ if there is an abupt stop before ending """
+        if self.progress_type == ProgressType.Determinate:
+            self.add(message='Interupted', value=0.0, total_val=1)
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: Inter**r**upted?

Also, having to check the progress_type everywhere to pass in the right values (especially when they are always the same for value and total_val) do not seem quite right to me. Is there a reason why they are not defaulted by the reporter?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 40]
```diff
@@ -0,0 +1,176 @@
...
+        percent = self.curr_val / self.total_val
+        return {'message': self.message, 'percent': percent}
+
+
+class InDetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self):
+        self.message = ''
+
+    def add(self, kwargs):
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: **kwargs?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 41]
```diff
@@ -0,0 +1,176 @@
...
+    def _format_value(msg, percent):
+        bar_len = BAR_LEN - len(msg) - 1
+        completed = int(bar_len * percent)
+
+        message = '\r{}['.format(msg)
+        message += ('#' * completed).ljust(bar_len)
+        message += ']  {:.4%}'.format(percent)
+        return message
+
+    def write(self, kwargs):
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: **kwargs?

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 42]
```diff
@@ -0,0 +1,176 @@
...
+    # for views that want to accept both determinate and indeterminate progres
+    Both = 2
+
+
+class ProgressViewBase(object):
+    """ a view base for progress reporting """
+    def __init__(self, out, progress_type, format_percent=None):
+        self.out = out
+        assert isinstance(progress_type, ProgressType)
+        self.progress_type = progress_type
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: It would be a worthy goal to avoid having to have this attribute...

File: src/azure-cli-core/azure/cli/core/commands/progress.py [hunk 43]
```diff
@@ -0,0 +1,176 @@
...
+
+class IndeterminateStandardOut(ProgressViewBase):
+    """ custom output for progress reporting """
+    def __init__(self, out=None):
+        super(IndeterminateStandardOut, self).__init__(
+            out if out else sys.stderr, ProgressType.Indeterminate, None)
+        self.spinner = humanfriendly.Spinner(label='In Progress')
+        self.spinner.hide_cursor = False
+
+    def write(self, kwargs):
```
Comment (johanste) [inline src/azure-cli-core/azure/cli/core/commands/progress.py]: **kwargs?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py (hunk 36 above)
Comment (oakeyc) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: because of cyclic imports

**Comment by oakeyc** [comment]
@troydai and @derekbekoe, can you please take another look

**Comment by troydai** [comment]
Reviewing.

File: src/command_modules/azure-cli-shell/azclishell/progress.py [hunk 44]
```diff
@@ -0,0 +1,95 @@
...
+                counter = 0
+                while counter < buffer_size:
+                    beat = HEART_BEAT_VALUES[_get_heart_frequency()]
+                    HEART_BEAT += beat
+                    counter += len(beat)
+        doc = u'{}:{}'.format(progress, HEART_BEAT)
+    shell.cli.buffers['progress'].reset(
+        initial_document=Document(doc))
+    shell.cli.request_redraw()
+    if PROGRESS == 'Finished' or PROGRESS == 'Interupted':
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-shell/azclishell/progress.py]: Typo 'Interupted'

Comment (oakeyc) [inline src/command_modules/azure-cli-shell/azclishell/progress.py]: Done
